src/main/java/org/jabref/gui/sidepane/SidePaneContainerViewModel.java

```diff
@@ -0,0 +1,159 @@
+
+    public SidePaneContainerViewModel(PreferencesService preferencesService) {
+        this.preferencesService = preferencesService;
+        visiblePanes.addListener((ListChangeListener<SidePaneType>) change -> {
+            while (change.next()) {
+                if (change.wasAdded()) {
+                    SidePaneType addedSidePane = change.getAddedSubList().get(0);
+                    switch (addedSidePane) {
+                        case GROUPS -> groupsPaneVisible.setValue(true);
+                        case OPEN_OFFICE -> openOfficePaneVisible.setValue(true);
+                        case WEB_SEARCH -> webSearchPaneVisible.setValue(true);
+                    }
+                } else if (change.wasRemoved()) {
+                    SidePaneType removedSidePane = change.getRemoved().get(0);
+                    switch (removedSidePane) {
+                        case GROUPS -> groupsPaneVisible.setValue(false);
+                        case OPEN_OFFICE -> openOfficePaneVisible.setValue(false);
+                        case WEB_SEARCH -> webSearchPaneVisible.setValue(false);
+                    }
+                }
+            }
+        });
+    }
+
+    /**
+     * Stores the current configuration of visible panes in the preferences,
+     * so that we show panes at the preferred position next time.
+     */
+    private void updatePreferredPositions() {
+        Map<SidePaneType, Integer> preferredPositions = new HashMap<>(preferencesService.getSidePanePreferences().getPreferredPositions());
+        IntStream.range(0, visiblePanes.size()).forEach(i -> preferredPositions.put(visiblePanes.get(i), i));
+        preferencesService.getSidePanePreferences().setPreferredPositions(preferredPositions);
+    }
+
+    public ObservableList<SidePaneType> getVisiblePanes() {
+        return visiblePanes;
+    }
+
+    /**
+     * @return True if <b>sidePane</b> is visible, and it can still move up therefore we should update the view
+     */
+    public boolean moveUp(SidePaneType sidePane) {
+        if (visiblePanes.contains(sidePane)) {
+            int currentPosition = visiblePanes.indexOf(sidePane);
+            if (currentPosition > 0) {
+                int newPosition = currentPosition - 1;
+                visiblePanes.remove(currentPosition);
+                visiblePanes.add(newPosition, sidePane);
+
+                updatePreferredPositions();
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * @return True if <b>sidePane</b> is visible, and it can still move down therefore we should update the view
+     */
+    public boolean moveDown(SidePaneType sidePane) {
+        if (visiblePanes.contains(sidePane)) {
+            int currentPosition = visiblePanes.indexOf(sidePane);
+            if (currentPosition < (visiblePanes.size() - 1)) {
+                int newPosition = currentPosition + 1;
+                visiblePanes.remove(currentPosition);
```

Comment:
Swapping elements in an ObservableLists is rather difficult to handle. The problem with this code here is that two events are fired (one for removing, and on for adding). Perhaps the easiest solution is to use the `sort` method from https://docs.oracle.com/javase/8/javafx/api/javafx/collections/FXCollections.html#sort-javafx.collections.ObservableList-java.util.Comparator- using a custom comparator. Since the list is small we can also be inefficient and do this as follows: create a copy of visiblePanes, swap the elements there and then use this new list as a reference for the comparator. Maybe you can mostly reuse the existing code with this strategy: instead of updating the visiblePanes directly, change the preferred position in the preferences and then use `sort(visiblePanes, ..)` using the PreferredIndexSort comparator.

Reply:
The second strategy is clever but not reusable, since both strategies are equally `≈` inefficient I settled with the first one, here is the result
```java
private <T> void swap(ObservableList<T> observableList, int i, int j) {
    List<T> placeholder = new ArrayList<>(observableList);
    Collections.swap(placeholder, i, j);
    observableList.sort(Comparator.comparingInt(placeholder::indexOf));
}
```